Check for datadir after the config files were read

Flowdalic opened the pull request:
If we first check for datadir existence and then read the config file
we bail out in situations where the default datadir is invalid,
e.g. because of an non-existing HOME, but the datadir specified in
config files is valid. We should not do that, as the user may specified a
valid datadir in the config file and expects the setting to take effect.

See for example https://github.com/bitcoin/bitcoin/pull/12255#issuecomment-403666092 where users hit this.

Flowdalic committed:
Check for datadir after the config files were read

If we first check for datadir existence and then read the config file
we bail out in situations where the default datadir is invalid,
e.g. because of an non-existing HOME, but the datadir specified in
config files is valid. We should not do that, as the user specified a
valid datadir and expects the setting to take effect.

Also remove datadir check in ReadConfigFiles() which now became
superfluous: We check for the datadir in AppInit() already, there is
no need to check also in ReadConfigFiles().

Furthermore re-enable the disabled tests in feature_config_args.py,
which where disabled in fabe28a0cdc ("qa: Temporarily disable test
that reads the default datadir location"), as those tests do not touch
"$HOME/.bitcoin" any more.

MarcoFalke commented:
Doesn't the location of the default config file depend on the datadir?

Flowdalic commented:
> Doesn't the location of the default config file depend on the datadir?

Does not appear to be the case: 

```
BITCOIN_CONF_FILENAME = "bitcoin.conf"
```

is used if not explicitly specified via `-conf`. Which would mean that it is opened relative to the working directory of the process. But even if the location of the default config file is dependent on the datadir, I don't see a problem.  If I'm not overlooking something.

MarcoFalke commented:
See https://dev.visucore.com/bitcoin/doxygen/class_args_manager.html#aec848fd54ace0df3df48dac446feb7d2 which calls `GetDataDir`

AtsukiTak commented:
Now the second if clause seems meaningless.

https://github.com/Flowdalic/bitcoin/blob/94a038b8694229c0153ede48a53701424ca51e38/src/bitcoind.cpp#L98-L102

Because it already checked in `ReadConfigFiles`.

https://github.com/bitcoin/bitcoin/blob/4a7e64fc85461a205f2b51da52d1455795d43b91/src/util.cpp#L909-L912

Flowdalic commented:
> Now the second if clause seems meaningless.

But it isn't: The datadir may have been set in a configuration file, hence we need to check after the config files have been read.

Flowdalic commented:
Right, the check can get probably removed.

Flowdalic committed:
Do not create datadir when calling GetConfigFile

As the real datadir may be set in the configuration file, this would
case the default datadir to get created.

Flowdalic committed:
Show the datadir that was checked in error message

Flowdalic committed:
Do not fallback to empy path in GetDataDir()

If datadir is not a directory, do not fallback to empty path as this would shadow the original path set in the config file. Consider

    bitcoin.conf:
    datadir=/does/not/exist

and bitcoind started with "-conf=bitcoin.conf", which would previusly yield

    Error: Specified data directory "" does not exist.

when it should state

    Error: Specified data directory "/does/not/exist" does not exist.

Flowdalic commented:
Rebased. Also fixed the cause and re-enabled the tests which were disabled in https://github.com/bitcoin/bitcoin/pull/13687 with fabe28a0cdcfa13e0e595a0905e3642a960d3077.

I had a hard time working on the argument and configuration file parsing code. It is a little bit spagettish and some parts are messy.

I used default argument values in order to optionally change GetDataDir() to not create the data directory, which, I assumed, made the change less intrusive. But I am happy to change it towards a GetDataDir() / GetDataDirAndCreateIfNecessary(). I am undecied which of those two approaches would be better.

Flowdalic commented:
The failed test `wallet_import_rescan.py` is green here. Could some restart anther travis run?

AtsukiTak commented on src/bitcoin-cli.cpp:
```diff
@@ -121,14 +121,14 @@ static int AppInitRPC(int argc, char* argv[])
         fprintf(stderr, "Error reading configuration file: %s\n", error.c_str());
         return EXIT_FAILURE;
     }
+    if (!fs::is_directory(GetDataDir(false))) {
```
It seems never fail unless user specify unpermitted path.

Flowdalic replied:
It fails if the specified path is not a directory. I think that is sensible and was already previous behavior. Or do have a specific change request?

AtsukiTak replied:
If I specify a file, `boost::filesystem::create_directory` throw an error.
If I specify a non-exist path, a new directory is created and no error happens.

I think you just missed second argument; `GetDataDir(false, false)` might be what you want.

AtsukiTak replied:
Sorry for my poor English. Please feel free to ask me if my English is not clear.

Flowdalic commented:
> I think you should leave only first one

Well, those commits depend on each other: For example I wouldn't be able to re-enable the relevant tests of "Check for datadir after the config files were read" without the previous commits.

I don't feel like that this patch series is worth splitting up, as it is pretty small, but I am happy create single PRs for each commit one after another, if there is a consensus that this would be favorable.

AtsukiTak commented:
> I don't feel like that this patch series is worth splitting up, as it is pretty small

I see. So please don't care about that. That was just newbie's nonsense comment.

Flowdalic committed:
Merge branch 'master' into init-swap-datadir-readconf

Flowdalic committed:
Merge branch 'master' into init-swap-datadir-readconf

leishman commented:
I don't think our PRs conflict, but just wanted to make you aware of a minor change I'm putting in the logging behavior: https://github.com/bitcoin/bitcoin/pull/14057

I'm also working on logic to create a `bitcoin.conf.example` file: https://github.com/bitcoin/bitcoin/pull/13761

Flowdalic committed:
Merge remote-tracking branch 'origin/master' into init-swap-datadir-readconf

Flowdalic commented:
Pinging @MarcoFalke since this PR re-enables tests he disabled in fabe28a0cdcfa13e0e595a0905e3642a960d3077. Would be great to hear your thoughts on this.

MarcoFalke commented:
I haven't looked at the most recent code, but I still think that the location of the default config file depends on the datadir? See https://dev.visucore.com/bitcoin/doxygen/class_args_manager.html#aec848fd54ace0df3df48dac446feb7d2 which calls `GetDataDir`.

So you are changing how and what config files are read. Not sure if we want this behavior change without discussion first.

MarcoFalke commented:
Also, you modified how bitcoind and bitcoin-cli work, but not the gui. Wouldn't that lead to issues when the gui is used as rpc server via the bitcoin-cli?

fanquake commented:
Closing "up for grabs".
